### ISSUE TITLE
Set cookies in request?

### ISSUE THREAD

**Issue by Jamesits** [Pathoschild/FluentHttpClient#72]
Is there any ways to manually set a specific cookie in request, or maybe a way to maintain a cookiejar across multiple requests?

If my tests are correct, `HTTPClient.GetAsync(...).WithHeader("Cookie", "A=B")` won't actually set cookie.

**Comment by Pathoschild** [comment]
The underlying `HttpClient` silently ignores cookies by default. It's a bit clunky, but you can override the `HttpClient` to avoid that. At that point you can call `.WithHeader("Cookie", "...")` and it'll work fine.
```cs
var client = new FluentClient("https://...", new HttpClient(new HttpClientHandler { UseCookies = false }));
```

(I'd love to add a `.WithCookie(...)` method to hide all that, but the complication is that we'd need to override the handler to do it. Since you can pass in your own client, it wouldn't work reliably.)

**Comment by Pathoschild** [comment]
I'll reopen this to fix cookie support by default. I won't add a `.WithCookie` method per my previous comment, but you'll be able to use `WithHeader("Cookie", "...")` by default.

**Comment by Jericho** [comment]
Here's a crazy idea: why don't we create a new fluent client capable of maintaining it's own cookie jar?

This class would not allow you to provide your own `HttpClient` and therefore would be able to hide the complication associated with cookies. This class would either derive from `FluentClient` or implement the `IClient` interface.

Maybe something like this:
```csharp
public class FluentClientWithCookies : IClient
{

}
```

What do you think?

**Comment by Pathoschild** [comment]
That's an interesting idea, but we can generalise it — there are other features we can add if we don't allow overriding the `HttpClient`.

But that brings up a question: what are the use cases which benefit from overriding the `HttpClient`? If it's only unit tests, maybe we can drop support for overriding `HttpClient` and add unit testability support directly? That would let us support features like cookies without a separate client.

**Comment by Jericho** [comment]
On my end, unit testing is the only reason why I override the httpclient but I allow the developers who use my library to provide their own client (which I pass to the fluent client constructor) so I would have to survey them to know if they would be ok with removing the ability to use a custom httpclient.

What other scenarios you think we could support?

**Comment by Pathoschild** [comment]
Some potential benefits to controlling the underlying client:

* Built-in cookie support:
  ```cs
  await client.GetAsync(...).WithCookie("key", "value");
  ```
* Built-in proxy support (we currently have a separate non-client-overriding constructor):
  ```cs
  var client = new FluentClient("...", new FluentClientOptions { Proxy = ... });
  ```
* Built-in support for [handler](https://msdn.microsoft.com/en-us/library/system.net.http.httpclienthandler.aspx) options:
  ```cs
  var client = new FluentClient("...", new FluentClientOptions { AllowAutoRedirect = true });
  ```
* Built-in unit testing:
  ```cs
  var client = new FluentClient(...).SetUnitTestMock(config => config
     .When(HttpMethod.Get, "*", (request, response) => response.Status(HttpStatusCode.NotFound))
     .When(HttpMethod.Post, "*", (request, response) => response.Body("...").Status(HttpStatusCode.OK));
  );
  ```
* We'd only need two constructors instead of 4+:
  ```cs
  public FluentClient(string baseUri, FluentClientOptions options = null);
  public FluentClient(Uri baseUri, FluentClientOptions options = null);
  ```

Some potential disadvantages:
* Can't directly use mocking libraries like [MockHttp](https://github.com/richardszalay/mockhttp).
* Can't support use cases that involve replacing the HttpClient or HttpClientHandler (if any)?

**Comment by Pathoschild** [comment]
Fixed in `develop` for the upcoming 3.2 release:

```cs
// set cookie for one request
client.GetAsync(...).WithHeader("Cookie", "...");

// set cookie for all requests
client.AddDefault(request => request.WithHeader("Cookie", "..."));
```
